fabtests/pytest/common.py

```diff
@@ -287,65 +290,89 @@ def prepare_base_command(self, command_type, executable,
 
         return command
 
+    def _run_client_command(self, server_process):
 
-    @retry(retry_on_exception=is_ssh_connection_error, stop_max_attempt_number=3, wait_fixed=5000)
+        if server_process.poll():
+            raise RuntimeError("Server has terminated")
+
+        print("")
+        print("client_command: " + self._client_command)
+
+        process = Popen(self._client_command, stdout=subprocess.PIPE,
+                        stderr=subprocess.STDOUT, shell=True, universal_newlines=True)
+        client_timed_out = False
+        output = ""
+        try:
+            output, _ = process.communicate(timeout=self._timeout)
+        except TimeoutExpired:
+            client_timed_out = True
+            process.terminate()
+
+        if has_ssh_connection_err_msg(output):
+            print("client encountered ssh connection issue!")
+            raise SshConnectionError()
+
+        print("client_stdout:")
+        print(output)
+        print(f"client returncode: {process.returncode}")
+
+        if client_timed_out:
+            raise RuntimeError("Client timed out")
+
+        return process.returncode
+
+    @retry(retry_on_exception=is_ssh_connection_error, stop_max_attempt_number=3, wait_fixed=SERVER_RESTART_DELAY_MS)
     def run(self):
-        if self._server_cmdline_args.is_test_excluded(self._server_base_command):
+        if self._cmdline_args.is_test_excluded(self._server_base_command):
             pytest.skip("excluded")
 
         if self._cmdline_args.is_test_excluded(self._client_base_command):
             pytest.skip("excluded")
 
-        # start running
-        server_outfile = NamedTemporaryFile(prefix="fabtests_server.out.").name
-        server_process = Popen(self._server_command + " > " + server_outfile + " 2>&1", shell=True)
+        # Start server
+        print("")
+        print("server_command: " + self._server_command)
+        server_process = Popen(self._server_command, stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT, shell=True, universal_newlines=True)
         sleep(1)
-        client_outfile = NamedTemporaryFile(prefix="fabtests_client.out.").name
-        client_process = Popen(self._client_command + " > " + client_outfile + " 2>&1", shell=True)
 
-        server_timed_out = False
+        client_returncode = -1
         try:
-            server_process.wait(timeout=self._timeout)
-        except TimeoutExpired:
+            # Start client
+            # Retry on SSH connection error until server timeout
+            client_returncode = retry(
```

Comment:
It seems you can put the `retry` decorator on `_run_client_command`, and call `self._run_client_command` directly here.

Reply:
Yes -  the reason I put it here is because of the use of `self._timeout` variable. Instance variables are not accessible in the decorator.

Reply:
```
stop_max_delay=self._timeout * 1000,
```
Is this value necessary? what does it do?

Reply:
In the new approach the client behavior is:
1. Check if the server is running. Terminate if not.
2. Try the client command.
3. In case of SSH error, start over again.
 
This all happens before we timeout the server, so we need to limit the max amount of time of retrying the client. `self._timeout` suits that purpose.

Reply:
What does `stop_max_delay` do?

Also, it seems the unit is in millisecond, that is why you multiply timeout by 1000?

Reply:
`stop_max_delay=T` stops the retry from `T` after the first attempt.

Yes - 1000 is the conversion from sec to ms.

Reply:
Thanks! I think a comment about converting second to ms would be helpful here.

Reply:
Thanks! Fixed in new push.